File: frontend/src/services/api/academic.ts
import { apiClient } from "./client"
import type { Faculty, Department, Programme, Course } from "@/types/academic"
import type { RegisterCoursesRequest, RegisterCoursesResponse } from "@/types/registration"

export const academicApi = {
  registerCourses: async (data: RegisterCoursesRequest): Promise<RegisterCoursesResponse> => {
    const res = await apiClient.post("/academic/registration/register", data)
    return res.data
  },

  createFaculty: async (data: { name: string; code: string; description?: string }) => {
    const res = await apiClient.post("/academic/faculties", data)
    return res.data
  },

  listFaculties: async (): Promise<Faculty[]> => {
    const res = await apiClient.get("/academic/faculties")
    return res.data
  },

  createDepartment: async (data: { faculty_id: string; name: string; code: string }) => {
    const res = await apiClient.post("/academic/departments", data)
    return res.data
  },

  listDepartments: async (facultyId: string): Promise<Department[]> => {
    const res = await apiClient.get(`/academic/departments/faculty/${facultyId}`)
    return res.data
  },

  createProgramme: async (data: {
    faculty_id: string
    department_id: string
    name: string
    code: string
    duration_years?: number
    capacity_planned?: number
    required_subjects?: string[]
    minimum_grades?: Record<string, string>
    aggregate_threshold?: number
  }) => {
    const res = await apiClient.post("/academic/programmes", data)
    return res.data
  },

  listProgrammes: async (): Promise<Programme[]> => {
    const res = await apiClient.get("/academic/programmes")
    return res.data
  },

  getProgramme: async (id: string): Promise<Programme> => {
    const res = await apiClient.get(`/academic/programmes/${id}`)
    return res.data
  },

  createCourse: async (data: {
    code: string
    name: string
    description?: string
    credit_hours: number
    course_type?: string
    prerequisites?: string[]
    lecturer_id?: string
  }) => {
    const res = await apiClient.post("/academic/courses", data)
    return res.data
  },

  listCourses: async (): Promise<Course[]> => {
    const res = await apiClient.get("/academic/courses")
    return res.data
  },

  createAcademicCalendar: async (data: {
    academic_year: string
    semester: string
    registration_open?: string
    registration_close?: string
    exam_period_start?: string
    exam_period_end?: string
  }) => {
    const res = await apiClient.post("/academic/calendar", data)
    return res.data
  },
}

File: backend/app/infrastructure/models/academic.py
from beanie import Document, Indexed
from pydantic import Field
from typing import Optional, List
from datetime import datetime

class Faculty(Document):
    tenant_id: str
    name: str
    code: str
    description: Optional[str] = None
    dean_id: Optional[str] = None

    class Settings:
        name = "faculties"

class Department(Document):
    tenant_id: str
    faculty_id: str
    name: str
    code: str
    description: Optional[str] = None
    head_id: Optional[str] = None

    class Settings:
        name = "departments"

class Program(Document):
    tenant_id: str
    department_id: str
    faculty_id: str
    name: str
    code: str
    description: Optional[str] = None
    duration_years: int = 4

    required_subjects: List[str] = []
    minimum_grades: dict = {}
    aggregate_threshold: Optional[int] = None

    accreditation_status: str = "accredited"
    accreditation_body: Optional[str] = None
    accreditation_expiry: Optional[datetime] = None

    capacity_planned: int = 100
    capacity_current: int = 100
    capacity_reserved: int = 0

    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "programs"
        indexes = [
            [("tenant_id", 1), ("code", 1)],
            [("department_id", 1)],
        ]

class Curriculum(Document):
    tenant_id: str
    program_id: str
    version: str
    effective_date: datetime
    end_date: Optional[datetime] = None

    courses_by_level: dict = {}

    class Settings:
        name = "curriculums"

class Course(Document):
    tenant_id: str
    code: Indexed(str)
    name: str
    description: Optional[str] = None
    credit_hours: int

    course_type: str = "core"

    prerequisites: List[str] = []
    corequisites: List[str] = []
    antirequisites: List[str] = []

    lecturer_id: Optional[str] = None

    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "courses"

class Accreditation(Document):
    tenant_id: str
    program_id: str
    body: str
    status: str
    issue_date: datetime
    expiry_date: datetime

    class Settings:
        name = "accreditations"


class Registration(Document):
    tenant_id: str
    student_id: str
    course_ids: List[str]
    academic_year: str
    semester: str
    total_credits: int
    status: str = "registered"

    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "registrations"
        indexes = [
            [("tenant_id", 1), ("student_id", 1)],
            [("tenant_id", 1), ("academic_year", 1), ("semester", 1)],
        ]


class AcademicCalendar(Document):
    tenant_id: str
    academic_year: str
    semester: str
    registration_open: Optional[datetime] = None
    registration_close: Optional[datetime] = None
    exam_period_start: Optional[datetime] = None
    exam_period_end: Optional[datetime] = None

    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "academic_calendars"
        indexes = [
            [("tenant_id", 1), ("academic_year", 1), ("semester", 1)],
        ]
